Add blog: Integrating Mailchimp with your Node.js App

yashovardhan opened the pull request:
Hey folks,

Please check out the blog you all had requested. I've marked this PR as WIP since I was looking to make some edits to the content and give a nice ending to the blog as well. All the technical parts of it are working.

Also, I wanted to ask you all about how the image integration works for your website specifically, do I need to specify the image size?

yashovardhan committed:
authors.ts: Add Yash as author

yashovardhan committed:
Add mailchimp-nodejs integration blog

evantahler commented:
Hi @yashovardhan! 

https://github.com/grouparoo/www.grouparoo.com/pull/511 should fix the failing tests on this branch.  We've enabled CI to run against forks, and the Github actions should run now as well to generate the sitemap, etc.  If you rebase your branch against `main` and push, the robots should get to work. 

@andyjih will get back to you soon about the Images!

yashovardhan commented:
I have updated the fork but it seems the CI is still failing. I'll try to fix the linting issues, though I'm not sure about the tests.

evantahler commented:
Ah! The failing test and the failing Github action are related - we have [this action](https://github.com/grouparoo/www.grouparoo.com/blob/main/.github/workflows/update-branch.yml) that rebuilds the sitemap on ever push, and the test checks to make sure that every page is in the sitemap.  Since we don't have write access to your fork, the action can't write the sitemap.

Easy fix though - just run `./scripts/github_action` to emulate what the Github action would do, and commit the changes to the sitemap.

andyjih commented:
@yashovardhan Hi Yash! To answer your questions: yes, the images do require a `height` and a `width`. They also need an `alt` text.

I'll add some ideas for a way to wrap up the blogpost as well! Thanks for this post!

andyjih commented:
@yashovardhan Overall the blogpost looks good! For ways to wrap up at the end, I had some ideas in the Google Doc I shared with you. Let me know if that might work for you, or we can try a different approach.

yashovardhan committed:
Update sitemap

yashovardhan committed:
Update integrating-mailchimp-with-your-nodejs-app.mdx

yashovardhan commented:
Updated the sitemap and fixed the images. 

Also made considerable grammatical changes, added a few instances to make the blog looked polished and added the grouparoo plug at the end.

yashovardhan committed:
Fix linting issues

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+---
+title: "Integrating MailChimp with your Node.js App"
```
["MailChimp" should be "Mailchimp"](https://mailchimp.com/about/brand-assets/)

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+---
+title: "Integrating MailChimp with your Node.js App"
+date: "2021-06-01"
```
Keep an eye on this, based on how long this PR hangs out there. You'll want to make sure it pops up to the top of the list when it is eventually published.

yashovardhan replied:
definitely! updated the date for today for the time being

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+# Integrating MailChimp with your Node.js App
+
+If you're reading this blog right now, the chances are that you're finding your way to add a MailChimp integration into your Node.js application. Don't worry; we have you covered! This is a step by step guide that will help you build a sample Node.js backend for your webpage and help you integrate it with the MailChimp API.
+This tutorial will discuss all the different things you need to look for within MailChimp with references to their excellent API documentation. We'll also talk about an alternative method using Grouparoo, which will do all the heavy lifting for you!
```
Not sure if you were trying to start a new paragraph here or not, but you'll want two newlines to create a new paragraph.

yashovardhan replied:
added

yashovardhan replied:
oh yeah, didn't notice it earlier. Thanks for pointing it out! Removing the heading.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+tags: [connections, engineering]
+---
+
+# Integrating MailChimp with your Node.js App
```
This is unnecessary, and we'll get dinged in SEO points for having it, as the title is already rendered as an `<h1>` at the top of the page. [See preview here](https://www-grouparoo-git-fork-yashovardhan-mailchimp-nodejs-gr-514bff.vercel.app/blog/integrating-mailchimp-with-your-nodejs-app).

yashovardhan replied:
thanks for pointing it out, I've removed the heading.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+# Integrating MailChimp with your Node.js App
+
+If you're reading this blog right now, the chances are that you're finding your way to add a MailChimp integration into your Node.js application. Don't worry; we have you covered! This is a step by step guide that will help you build a sample Node.js backend for your webpage and help you integrate it with the MailChimp API.
+This tutorial will discuss all the different things you need to look for within MailChimp with references to their excellent API documentation. We'll also talk about an alternative method using Grouparoo, which will do all the heavy lifting for you!
```
I would probably remove the "if you're reading this ..." piece. Presenting the high-level view of what you're going to talk about is a fine enough introduction in a more technical piece like this.

Perhaps something like this?

> This is a step-by-step guide that will help you integrate Mailchimp with your Node.js application using Mailchimp's API. We'll begin by walking through the process manually, and end by showing you an easier approach that lets Grouparoo do all the heavy lifting for you.

yashovardhan replied:
makes sense, changing the initial start paragraph.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+
+- A MailChimp Account
+- Node.js & npm installed on your machine
+- A basic frontend application to send requests to your Node.js application.
```
Links here would be helpful. Mailchimp is self-explanatory, but I usually like to have a reference to what I'm asking readers to accomplish. In this case, links to docs on how to install Node and NPM would be helpful.

For the front end app, that could be intimidating to some folks. Maybe they already have a front end app and that's okay. If not, it might be nice to point to something you'd recommend starting with, like Next.js.

yashovardhan replied:
I'm including a very simple frontend example from w3schools.

yashovardhan replied:
added links for everything.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+- Node.js & npm installed on your machine
+- A basic frontend application to send requests to your Node.js application.
+
+This guide will help you make a sample Node.js application with Express to send and receive API requests and responses. Although a basic knowledge of Node.js would be helpful, you can easily follow the steps given and get started with your application.
```
When paired with the previous comment, I'm starting to feel like you're asking the reader to do a lot without a direct reference.

If we're going super bare-bones here, which it looks like you are, I'm wondering if you should provide an example repo. You could include both the front and the back end. And if you build it out incrementally, you could point to specific points in time using git commit hashes on GitHub. And in doing so, you're getting started section becomes way simpler. You point to a README this example repo that asks you to clone, install dependencies, and then provides two commands — one to run the front end, one to run the back end.

yashovardhan replied:
I feel the w3schools sample frontend will be helpful here.

I'm rather on the fence on providing the whole code as a GitHub repository since people might tend to copy that instead of reading through this and preferring grouparoo as a much better option.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+
+## Setting up your MailChimp account
+
+Setting up your MailChimp account is a relatively simple process. In this tutorial, we're assuming that you have created a free MailChimp account and are looking to integrate it within your application. The process is pretty similar for premium accounts, just that you can create multiple audiences, each having their separate Audience ID. To integrate your app with Mailchimp, we need to pieces of  ...
```
_two_ pieces?

yashovardhan replied:
oops, fixed.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+  alt="MailChimp Audience ID"
+  width={483}
+  height={116}
+/>
```
I'm on the fence about the value of these sections. On one hand, it's nice that you handhold the reader through the process. On the other hand, these steps could change tomorrow when Mailchimp decides they want to make a change to their application, and then this is out of date. And [Mailchimp has well-documented info around API keys](https://mailchimp.com/help/about-api-keys/) and [finding audience ID](https://mailchimp.com/help/find-audience-id/).

My gut is to remove them and point to the documentation. Then you can spend more space on the details of the application and less on the setup.

yashovardhan replied:
I believe this is a total beginners' guide so the handholding is much needed. I want the users to directly see this blog rather than anything else in the Mailchimp documentation If they're looking to integrate it with the Node.js app.

The aim of this blog was to capture SEO as I remember. If we're redirecting folks to Mailchimp documentation it will be worse for us generally I believe. There are already existing blogs that deep dive into the application perspective and even Mailchimp has excellent documentation of building this rather.

Also, @andyjih has this part in his sample blog.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+
+Here, on click, the function checks for the validity of email and accordingly places a `POST request` on the '/subscribe' endpoint of our application with the body containing the `email` value entered by the user.
+
+For simplicity, we're logging in all the error messages into the browser console.
```
This is where I think it could be super valuable to be able to see it in action. To have a simple HTML and JS file to reference.

I also think this should be in another section about setting up the front end, not under _setting up your Node app_.

yashovardhan replied:
I can add that section, setting up your frontend to work with the node js backend.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
````diff
@@ -0,0 +1,350 @@
+
+It is time to make a `package.json` file to capture all the necessary modules required by our project in your repository.
+
+```
+$ npm init
+```
+
+This will ask you for a few project details and initiate a package.json file inside your repository.
````
I often suggest `npm init -y` because then it gets out of the way and we can move on with the details that matter to the project.

yashovardhan replied:
I tend to use just init rather as it lets me customise things initially.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+
+This will ask you for a few project details and initiate a package.json file inside your repository.
+
+Now it is time to install all our necessary modules for the project. We will require the following modules:
```
Subheadings here could be valuable. 

- Add package.json
- Install dependencies
- And so on ...

yashovardhan replied:
makes sense, I'll add that.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
````diff
@@ -0,0 +1,350 @@
+
+```
+app.use(express.json());
+```
+
+2. To parse the data that is coming from the POST request, we will be using `body-parser`
+
+```
+app.use(bodyParser.urlencoded({extended: false}));
+```
+
+Note that we have set `extended` to `false`, which means that the values can only be `strings` or `arrays` which is what we need for now.
+
+3. Next up, we need to specify the folder which contains your frontend application and connect it to the express server. Here, we’re having a folder named `public`, which includes all our frontend code.
+
+```
+app.use(express.static(path.join(__dirname, 'public')))
+```
+
+4. Perfect, let's create our `POST` request quickly. We're just setting it up to add a `console.log()` to check if everything's working as expected.
+
+```
+app.post('/subscribe', (req, res) => {
+    console.log(“I am listening”);
+})
+```
+
+5. Now let's define the `PORT` which will host our application.
+
+```
+const PORT = process.env.PORT || 8888;
+```
+
+This sets the value to the environment `PORT` value in case we're using a cloud hosting or default it to 8888, which is our local machine endpoint.
+
+6. Finally, for all this to work, let's start a socket that listens for our connections on the `PORT`.
+
+```
+app.listen(PORT, console.log('The server has started!'));
+```
````
This is tough to read in this way. I get that you're explaining one line at a time, but it's easy to get lost when I don't have the context of the full picture.

What I like to do in these scenarios is share a chunk of code, along with comments. And if anything requires more commentary than a single line, it's probably worth calling out _after_ sharing the snippet.

yashovardhan replied:
I tried that but it wasn't looking very easy for a beginner to understand. To make it much clearer I can write the code towards the end. What say?

seancdavis replied:
It's not _where_ you're presenting it that makes it challenging, it's _how_ you're doing it. I understand that a large file can look intimidating, but I don't think that's what we're talking about here.

I think the information you've presented here is great. It's just missing context of where it fits in the larger file. Without some reference to how it all looks together, it's easy to get lost.

yashovardhan replied:
I'll look into it and add something easier to read in that case

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+
+For this whole process, we’re directly referring to the [MailChimp API Documentation](https://mailchimp.com/developer/marketing/api/list-activity/). They have excellent documentation stating all the different things you need to be aware of with extremely detailed resources. Make sure to check that out if you need any additional information that is not mentioned here directly.
+
+We’re using the standard URL route here. You can also download the [MailChimp NodeJS module](https://www.npmjs.com/package/mailchimp-api-v3) if you want to as well. Although this makes your life easier in implementing the API, it adds many other modules that we do not need for this particular project.
```
I don't know that I buy this. If I were writing this, I would have done it the other way around — show the user the easier path (using a library that is well-supported) with some commentary on being cognizant of bringing in packages that you're only using to a minimal extent.

yashovardhan replied:
I get that you might prefer using the Mailchimp package rather, but it won't be very fruitful in terms of writing this blog. There's the whole Mailchimp package that does everything from managing the marketing campaigns to setting up the audience at a higher level. Here we're just going through a simple integration of a pre-existing node.js app. 

For a larger project, using the package might make sense, but for smaller projects, I'll prefer the API route.

I can rewrite my explanation here, and change it to something like:
```
We're using the Mailchimp API v3 here with the standard URL route. You can also download the [MailChimp NodeJS module](https://www.npmjs.com/package/mailchimp-api-v3) if you want, but that is preferable for larger projects rather than a small integration we're building here today.
```
```

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+
+### Getting your API Endpoint
+
+So the first thing here is to get the URL of your API. If you check the documentation, the URL looks something like `https://<dc>.api.mailchimp.com/3.0`
+
+Here the `<dc>` corresponds to the data center applicable for your account. If you go back to _step 4_ of **Getting your API Key**, you can see that for us, the data center is `us6`. Hence for us, the URL looks something like `https://us6.api.mailchimp.com/3.0`
+
```
This feels ... not right. You're asking the user to go back to earlier in the post and grab another value. At the very least this should read linearly so that you're either telling them what to do here or made note of the value when the user was following step 4.

But even making those adjustments feels a little weird. If you use node-mailchimp, you don't need to know this value. You can rely on the library to manage that for you.

yashovardhan replied:
I feel using the library is not a smart choice for such a basic integration blog. If we were diving deeper into accessing different aspects of Mailchimp using their module, it might have made a lot of sense.

For this step, I have clearly mentioned in the section of getting your API Key that you need to be mindful of this value. I've also repeated myself here so that anyone who's reading this might not have to go back and forth.

seancdavis commented on pages/blog/integrating-mailchimp-with-your-nodejs-app.mdx:
```diff
@@ -0,0 +1,350 @@
+
+As you can see, just to configure a single API function, we needed to go through a considerable integration pain, which again is a code you need to maintain regularly. What I generally use and recommend to you all is definitely, Grouparoo.
+
+Grouparoo is a data syncing platform that connects directly to your product database, allows you and your team to define what customer data you care about, and then syncs that data to Mailchimp. The best part, the tool is fully open source! You can easily choose between a self-hosted or a cloud offering. Using this will be just a matter of minutes to build this integration, and anyone can easily  ...
```
We have an example app that [connects SQLite data to Mailchimp](https://github.com/grouparoo/app-example-config). Might be useful to link to it here.

seancdavis reviewed (changes requested):
I really like this! I think it is well-structured and tells a nice story. I made a lot of comments that I think could bring this thing to the next level. The big ones are:

1. Provide the sample code. I think that would make it easier for the user to follow along.
2. No need to cover content that Mailchimp has in their documentation.
3. node-mailchimp package could really simplify the integration section here.

Let me know what you think or if you'd like me to clarify any of these thoughts.

yashovardhan committed:
Update blog with initial changes

yashovardhan commented:
Thanks a lot for your review @seancdavis. I've updated the blog and added some of my thoughts over in the comments as well.

Overall, I feel:
1. Providing a sample code might backfire on the main purpose of this blog.
2. If we're redirecting folks for setting up their Mailchimp account, we will be losing them for the integration aspect as well largely. This information will come in handy for them while thinking of using Grouparoo as well.
3. Although the Mailchimp module might simplify the process, using it for such a basic integration does not make a lot of sense for me rather.

The only other thing I wanted to add here is that this blog is focusing on the integration part (the name is integrating Mailchimp with your nodejs app) and I don't think it will be fruitful to just show how their module works.

seancdavis commented:
Overall I feel like you have a good article, but that these adjustments will really take it to the next level. Here are my responses:

> 1. Providing a sample code might backfire on the main purpose of this blog.

Disagree. I've had several readers ask for in-context code samples over the years. I can't recall a time where the extra context confused someone. By providing an example, you also have the benefit of longevity in that you can show which versions of libraries you're working with, which you lose without the broader context.

If you're worried about having another repo, this could be as simple as:

- GitHub Gist
- Showing the code examples more holistically in the post

> 2. If we're redirecting folks for setting up their Mailchimp account, we will be losing them for the integration aspect as well largely. This information will come in handy for them while thinking of using Grouparoo as well.

I'll buy the argument that keeping them focused on reading in one location is preferable. I still think it presents risk — this will likely become outdated before the code. I don't buy that someone is going to use Grouparoo because we taught them how to get an API key from Mailchimp. IOW I think it's fine to take on the risk in this case if you feel strongly about it. I'm on the fence about this one. We wouldn't put this information in our docs, but maybe it's okay in a blog post.

> 3. Although the Mailchimp module might simplify the process, using it for such a basic integration does not make a lot of sense for me rather.

I'm still failing to understand _why_ you don't think it makes sense, especially if it simplifies the process. I know the package does a lot of things that aren't necessary, but why does that matter in this case? Not using it means you're rewriting code that has already been written and tested (presumably).

FWIW [Grouparoo is using the package](https://github.com/grouparoo/grouparoo/blob/434bc7da56a6bce5e03b0e904cfc41f722f12026/plugins/%40grouparoo/mailchimp/package.json#L32).

yashovardhan commented:
> If you're worried about having another repo, this could be as simple as:
> 
> GitHub Gist
> Showing the code examples more holistically in the post

I like this approach a lot. I'll see how much code I can jam into the blog before it looks intimidating and rest I will use a GitHub gist to show the end to end code.

>I'll buy the argument that keeping them focused on reading in one location is preferable. I still think it presents risk — this will likely become outdated before the code. I don't buy that someone is going to use Grouparoo because we taught them how to get an API key from Mailchimp. IOW I think it's fine to take on the risk in this case if you feel strongly about it. I'm on the fence about this one. We wouldn't put this information in our docs, but maybe it's okay in a blog post.

For the Docs I agree it will not make sense. For a blog, I feel it can. I will additionally include the link to make sure people can go and check in case the graphics have changed.

>I'm still failing to understand why you don't think it makes sense, especially if it simplifies the process. I know the package does a lot of things that aren't necessary, but why does that matter in this case? Not using it means you're rewriting code that has already been written and tested (presumably).

I feel when *integration* is the main thing we're showing in a blog, just showing how a module works is not the best thing to show. If you feel so strongly about it, can we instead add another section here? Integrating using Mailchimp module? The current section will become integrating using the API method. This makes the blog have much more information in terms of integration, which is as I mentioned the key aspect of this blog.


Let me know what you think and I can accordingly start working on the blog.

seancdavis commented:
@yashovardhan:

> I'll see how much code I can jam into the blog before it looks intimidating and rest I will use a GitHub gist to show the end to end code.

👍 

> For the Docs I agree it will not make sense. For a blog, I feel it can. I will additionally include the link to make sure people can go and check in case the graphics have changed.

Love it! 👍 

> I feel when integration is the main thing we're showing in a blog, just showing how a module works is not the best thing to show. If you feel so strongly about it, can we instead add another section here? Integrating using Mailchimp module? The current section will become integrating using the API method. This makes the blog have much more information in terms of integration, which is as I mentioned the key aspect of this blog.

I don't think I'd add another section. Since you feel strongly about your approach, what you have is okay with me, especially if you have the larger example to reference to show how it all fits together.

andyjih commented:
@yashovardhan Any chance you can re-run `./scripts/github_action` and re-commit/push to your branch? Thanks!

yashovardhan committed:
Update blog with requested changes

yashovardhan commented:
@andyjih updated the blog as per requested changes. Here's the gist file associated with the project: https://gist.github.com/yashovardhan/6ac14edc99981dcc14d79627abb72cf8

yashovardhan committed:
MailChimp -> Mailchimp

yashovardhan committed:
Fix CI Issues

yashovardhan commented:
I'm not sure why the test is failing, can anyone help me here?

evantahler commented:
That failing test (while not giving the best possible error message) is saying that somewhere on the page there's an `<a>` link that's missing a `href`.  Check out the line below. 

```
Checkout this [example app]()https://github.com/grouparoo/app-example-config that connects SQLite data to Mailchimp.
```
The link should be within the `()`

yashovardhan committed:
Fix URL error

yashovardhan committed:
update sitemap

yashovardhan commented:
> That failing test (while not giving the best possible error message) is saying that somewhere on the page there's an `<a>` link that's missing a `href`. Check out the line below.
> 
> ```
> Checkout this [example app]()https://github.com/grouparoo/app-example-config that connects SQLite data to Mailchimp.
> ```
> 
> The link should be within the `()`

Thank you for helping out! I updated the blog 😄
